### ISSUE TITLE
QPDF Flatten Annotations AND Hyperlinks in a PDF file

### ISSUE THREAD

**Issue by Geo-Van** [qpdf/qpdf#1039]
Many many thanks for the great tool provided QPDF.
Please let me ask regarding the wonderful feature --flatten-annotations :

(1) Applying  **{ --flatten-annotations=all or --flatten-annotations=print or --flatten-annotations=screen }** on a PDF file,
it flattens ANNOTATIONS, but also it flattens all the HYPERLINKS contained in the initial pdf file.
Is there a way to flatten annotations but keep HYPERLINKS alive, please?

(2) Is there a way to flatten TEXT, please?
I mean to convert the text contained in a pdf file from searchable to not searchable. (something like Ghostscript command -dNoOutputFonts)

(3) Processes made with QPDF (for example -flatten-annotations, or --linearize, etc) are COMPLETELY PERMANENT (irreversible)?
Or is there a way to be reversed in any way?

**Comment by jberkenbilt** [comment]
1. Flattening annotations and keeping hyperlinks seems like correct behavior. I'll look into this. I think if it is not doing that, it was an oversight.
2. It is less likely that I would build this into the qpdf CLI as this is more content than structural, and making it work would vary across different ways of representing text. It needs a deeper PDF interpreter than qpdf. It would be possible to use qpdf to write a tool that would remove character maps from subsetted fonts, but this still wouldn't solve it for built-in fonts or fonts that don't remap their characters.
3. Changes made by qpdf are not reversible per se, but some changes don't result in any semantic changes, only structural changes. Linearization is not reversible, but you can use qpdf to write a non-linearized version of a linearized file. qpdf will never be able to reconstruct to the original file. Flattening annotations loses information from the original file, so that is not reversible. The intention of flattening annotations was originally to prepare files for printing.

I will leave this open to investigate your item 1. Thanks for the questions and for raising the issue.

**Comment by Geo-Van** [comment]
Thank you very much!
Please let me clarify, regarding:

(1) It would be wonderful to "add" a switch to keep or not keep hyperlinks when applying --flatten-annotations , please.

(2) Very clear, QPDF will always keep text searchable, there is not an option to make it unsearchable, thanks!

(3) 
a)Please let me know, how can we use qpdf to write a non-linearized version of a linearized file? What command can we use?
b)If i understand correct from your answer, ALL processes are irreversible meaning that will never retain "DATA" from the 
   original file in order to can use it later and come back to exactly the original file. Am i understand it correct, please?
c)Example ---> removing metadata using the command:        **_qpdf --empty --pages input.pdf -- output.pdf_**
  After applying this command  ALL pdf metadata are gone. Is there a way to get them back in any way, please?

Thank you very much!

**Comment by jberkenbilt** [comment]
To remove linearization, simply run qpdf without the `--linearize` flag.

Regarding 3, I guess it depends on what you mean by reversible. With no flags, qpdf is non-destructive. `qpdf in.pdf out.pdf` generates a PDF that is exactly equivalent semantically with the original, except that any digital signatures will be invalidated because digital signatures are dependent on the exact bytes. A PDF file is basically a bunch of objects (maps, lists, strings, etc.) with cross references to each other. It's conceptually similar to a format like JSON. In fact, qpdf has its own JSON format that is semantically equivalent to JSON. If you had a JSON file like this:
```json
{
  "one": 1,
  "two": 2,
  "three": 3
}
```
you could rewrite it as `{"one":1,"three":3,"two":2}`, and it would represent exactly the same information. Given either format, any JSON parser would get exactly the same information, but there would be no way to reconstruct the exact bytes of the first format from the second without knowledge of all the parameters about the way it was written. It is very much like this for qpdf and PDF. qpdf may write the output PDF file with the objects numbered different, in a different order, with different whitespace, etc., but all the information, including what I would call "metadata", will be preserved. qpdf has its own way of formatting objects and its own way of ordering them, but any compliant PDF reader doesn't care about those details.

Linearization is a separate process that imposes additional constraints upon ordering. Two linearized files may both follow those constraints but may be different within those constraints.

When a file is encrypted, there are passwords and encryption keys. If qpdf preserves encryption, which is the default, the same passwords and keys will be used, but the signatures will be invalidated. If PKI is used, qpdf can't preserve that without the private key, and at this time, qpdf doesn't have support for PKI-based signatures, though it is something I will add at some point if possible.

I hope this clarifies things.  If you'd like to dig deeper, you can check out by blog post on the structure of a PDF file. https://medium.com/@jberkenbilt/the-structure-of-a-pdf-file-6f08114a58f6

**Comment by Geo-Van** [comment]
Thank you very very much!!!
This explains everything.

So, because i am not specialized in pdf files structure, just to understand simple my 3rd question:

i) Using qpdf with the command **_--flatten-annotations=all_**
will flatten annotations and if we want to get back those flatten annotations, there is no way.
Do i understand it correct, please?

ii) Using qpdf  to metadata using the command: **_qpdf --empty --pages input.pdf -- output.pdf_**
will remove all metadata info (including XMP metadata), and if we want to get back those metadata info there is no way.
Do i understand it correct, please?

iii) We read regarding the ExifTool and QPDF the following:
https://cyberrunner.medium.com/removing-metadata-from-pdf-files-using-exiftool-and-qpdf-20090b75d7f0
Exiftool can be used to remove all the metadata, in which case the metadata fields are replaced with nulls.
**exiftool -all:all= CLEAN_file.pdf -overwrite_original**
The main disadvantage of edits with Exiftool is that the edits are reversible. Some else can recover the metadata as they are actually never deleted, as follows
**exiftool -pdf-update:all= ORIGINAL_file.pdf**
After using the Exiftool for editing, the tool qpdf can be used to linearize the pdf file in order to remove all the old metadata (which were replaced by nulls) to make the edits with Exiftool irreversible.
The tool qpdf can be used to remove all unreferenced (old metadata) objects, that were removed with Exiftool to make it permanent This prevents that the tags cannot be restored.
**qpdf --linearize in.pdf out.pdf**
SO:
The TWO commands   (1) exiftool -all:all= CLEAN_file.pdf -overwrite_original      AND  (2) qpdf --linearize in.pdf out.pdf
do have exactly same result with only ONE qpdf command ------> the following:   qpdf --empty --pages input.pdf -- output.pdf
Do i understand it correct, please?

We are looking forward for your confirmation regarding (i), (ii), and (iii) above if we get the concept correct, please.
We really appreciate your great help.

**Comment by jberkenbilt** [comment]
> i) Using qpdf with the command **_--flatten-annotations=all_** will flatten annotations and if we want to get back those flatten annotations, there is no way. Do i understand it correct, please?

Correct.

> ii) Using qpdf to remove all metadata from pdf file using the command: **_qpdf --empty --pages input.pdf -- output.pdf_** will remove all metadata info (including XMP metadata), and if we want to get back those metadata info there is no way. Do i understand it correct, please?

Correct.

For your third point, I can't say whether they are exactly the same since I don't know exactly what exiftool does. It's quite likely that the `qpdf --empty --pages input.pdf -- output.pdf` will discard different information and most likely more information. For example, for now, that will discard outlines (bookmarks), hyperlinks, and other document-level features -- a shortcoming I intend to address soon. Also, `--linearize` is not necessary to get the affect with qpdf. It would be possible to use qpdf to irreversibly remove metadata, but there is not a simple way. You could do it by manipulating the file using qpdf json. If exiftool removes metadata by replacing the reference with null and leaving the metadata unreferenced in the file, then running subsequently through qpdf will remove it with or without `--linearize`. The best way to be sure would be to check for it.

**Comment by Geo-Van** [comment]
Please let me discuss your last comment:

1.
**Your first point** : _For your third point, I can't say whether they are exactly the same since I don't know exactly what exiftool does. It's quite likely that the qpdf --empty --pages input.pdf -- output.pdf will discard different information and most likely more information. For example, for now, that will discard outlines (bookmarks), hyperlinks, and other document-level features -- a shortcoming I intend to address soon._ 
**Please let me discuss my findings:**
Using qpdf --empty --pages input.pdf -- output.pdf  discards NOTHING - only metadata. Outlines (bookmarks), hyperlinks, and other document-level features, are stll there. Only metadata is gone. This is good.

2.
**Your second point** :
_It would be possible to use qpdf to irreversibly remove metadata, but there is not a simple way. You could do it by manipulating the file using qpdf json._ 
**Please let me discuss my findings:**
Using qpdf --empty --pages input.pdf -- output.pdf  removes all metadata. This is good.
I do not understand why you mention that "but there is not a simple way. You could do it by manipulating the file using qpdf json." ?

3.
**Your third point** :
_If exiftool removes metadata by replacing the reference with null and leaving the metadata unreferenced in the file, then running subsequently through qpdf will remove it with or without --linearize. The best way to be sure would be to check for it._
**Please let me discuss my findings:**
YES you are correct - running subsequently through qpdf will remove it with or without --linearize. Thank you very much for let me know this.

Looking forward for your comments on 1 and 2 above. 
Thank you very much!

**Comment by jberkenbilt** [comment]
1. I have no explanation. When you start with `--empty`, as of the latest qpdf (currently 11.6.1), and for a long time, the only document-level feature that is copied is page labels. Perhaps we are differing on some terminology.
2. Typically what I would call "metadata" would be the XMP metadata, an embedded XML block inside most PDF files. qpdf treats this specially with encrypted files, but otherwise it does not. Since metadata is linked from the info and not through the document pages, it would typically not be preserved by `--pages` starting with empty, but it's possible that it could be if there are links to it from elsewhere in the document. For example, the PDF 1.4 spec has /Metadata that is not referenced elsewhere. This qpdf command strips it. The PDF 1.7 spec also links to the metadata from an annotation. That gets preserved. Maybe something similar is happening with outlines in your example file.

If I had to take a guess on exiftool, it would be manipulating xmp metadata, which is often (but not always) structured in a way that allows non-pdf-aware tools to modify it.

The bottom line is that qpdf doesn't handle these things in a special way at this time (though it is coming for a handful of document features). The `--pages` options copy page objects and anything they refer to, stopping at page boundaries. So any object in the PDF file that is not linked to by a page is lost, and anything else is preserved. Since a PDF file can be structured in an arbitrary way, for the most part, all I can say is what qpdf would do with a typical PDF file, but in this space, there will be exceptions.

**Comment by Geo-Van** [comment]
Thank you so much for the deep explanation.

As my FINAL conclusion according with all your comments is the following:

To remove ALL ALL ALL metadata (including XMP, annotation metadata, etc), the correct command is the following:
**qpdf --flatten-annotations=all --empty --pages input.pdf -- output.pdf** 

Am i understand it correct this time, please?

**Comment by jberkenbilt** [comment]
This will work correctly for most typical PDF files. It would be possible to construct a PDF file for which this would not work, but you wouldn't be that likely to encounter one. So I would say, "Yes with rare exceptions" would be the best answer. I would also note that, in some future version of qpdf hopefully within the next several months, additional flags may be required to intentionally strip some things that qpdf strips today. I haven't decided about that. If I change the default behavior, the qpdf version will definitely be >= 12.0.0, since I never change default behavior without a major version bump. I may leave the default behavior as it is and add extra flags. Just be sure to take a look at the release notes. Anyway, this issue has raised some awareness in my mind, and I will keep it in mind.

I will also keep the issue open since it has the suggestion about preserving hyperlinks when flattening annotations.

**Comment by Geo-Van** [comment]
Hello,
Thank you!

To eliminate to 0 the possibility of not work, i will use TWO commands:
**exiftool -all:all= input.pdf -overwrite_original** 
**qpdf --linearize --flatten-annotations=all --empty --pages input.pdf -- output.pdf**
I think with this method will eliminate to  0 the possibilities to not work, and will ensure 100% working remove metadata.
Are you agree with me, please?


Regarding the extra flags on qpdf ---> we will be very happy if in future version of qpdf we can 
use **--flatten-annotations=all**  with EXTRA FLAG to choose between :
- flatten ANNOTATIONS and flatten Hyperlinks
OR
- flatten ANNOTATIONS and NOT flatten Hyperlinks (keep hyperlinks alive)

**Comment by jberkenbilt** [comment]
I can't give you a 100% guarantee, but all that looks good, and I think throwing ghostscript into the mix is probably a good idea. You don't need `--linearize` for any of this to work. I can't promise that there isn't anything you forgot, but nothing is jumping out at me. And noted about the hyperlinks.

**Comment by Geo-Van** [comment]
Thank you very very much!

**Comment by Geo-Van** [comment]
Please just let me ask:
Using the command: **qpdf --empty --pages input.pdf -- output.pdf**
Is there a way to  use the **--replace-input** together with that command?
I try **qpdf --empty --pages input.pdf --replace-input  input.pdf**
but unfortunately is not working.....
The idea is to use that command ithout create a NEW pdf file but just replace it.

**Comment by jberkenbilt** [comment]
There's no way to use --replace-input with --empty because --empty is the input. The other files named in --pages are not considered inputs. All --replace-input does is write to a temporary file and rename the temporary file over the original if successful. I'm afraid that, for this case, you'll have to manage that yourself.

**Comment by Geo-Van** [comment]
Thank you!

In other cases  _**--replace-input**_   can be used because the **input file is CONVERTED on the FLY** to an output file by the help of a temporary file, and when finished the temporary file is AUTOMATICALLY erased so finally NOTHING is stored in the computer.
BUT in the case of  **_--empty_**   we can not use  _**--replace-input**_ , because it needs to write to a NEW CREATED FILE from the old input pdf file.

Do we get it correct, please?

**Comment by jberkenbilt** [comment]
Note to self: the remaining work here is for flatten annotations to not break hyperlinks.

**Comment by jberkenbilt** [comment]
Rereading this part of the spec, I think the correct behavior is to keep annotations that have no appearance streams. Per the spec, this is Link, Popup, and Projection, but I think qpdf should just keep the annotation if it has no appearance stream.

**Comment by jberkenbilt** [comment]
This was a trivial fix. Too bad it missed 11.7.0. It will be in whatever's next.

**Comment by Geo-Van** [comment]
Hello,
Thank you for your email.
So:

OLD BEHAVIOR (Before version 11.7.0):
Applying { --flatten-annotations=all or --flatten-annotations=print or --flatten-annotations=screen } on a PDF file,
it flattens ANNOTATIONS, but also it flattens all the HYPERLINKS contained in the initial pdf file.


NEW BEHAVIOR (After version 11.7.0):
Applying { --flatten-annotations=all or --flatten-annotations=print or --flatten-annotations=screen } on a PDF file,
it will flattens ANNOTATIONS, but will KEEP the HYPERLINKS contained in the initial pdf file.


BUT, please let me ask:
Can you add a NEW SWITCH command in order to can choose between OLD and NEW behavior, please?
(In other words to can choose FLATTEN all included the hyperlinks, OR FLATTEN all but keep alive the hyperlinks)

I am looking forward for your reply
Best Regards
George

________________________________

**Comment by jberkenbilt** [comment]
The new behavior is not in 11.7.0. It will be in 11.7.1 or 11.8.0, whatever the next release is. I have no intention of supporting the old behavior, which was a bug. The intention of flatten annotations was to take annotations that have separate appearances that are not part of the page description and add them to the page description. Annotations that had no appearance were not supposed to be touched. In most cases, qpdf was doing this correctly, but the logic was slightly wrong and was unintentionally dropping link, popup, and projection annotation types which have no appearance _by design_. Given that flattening link annotations would have no impact on the appearance of the page (print or display), I can see no real benefit to touching them as part of flattening annotations. Bear in mind that the all the link annotation does is make that area of the page active if you click on it. Any appearance that shows it to be a link is drawn on the page and is not part of the annotation.

**Comment by Geo-Van** [comment]
Hello,
Thank you very much for your email.
OK, i understand that this is the "correct" behavior for FLATTENING pdf document.

There is also another process regarding pdf documents named --> RASTERIZE
This process as known converts the pdf document to a pdf which is like an image - there are NO hyperlinks etc.
Can you please add to QPDF this wonderful feature, please?

Best Regards
George
________________________________

**Comment by jberkenbilt** [comment]
qpdf will never be able to rasterize a PDF -- that requires the ability to render, which is not something qpdf will do. qpdf doesn't have any ability to interpret the content stream of a page (other than in very limited cases) and doesn't have any awareness of graphics, fonts, etc. The usual tool for this purpose is ghostscript. Most other open source tools that rasterize PDF, such as ImageMagick, use ghostscript. There's also poppler and several other tools that can render PDF. If you wanted to use qpdf to disable hyperlinks, it would be straightforward to do it using the library, pikepdf, or even qpdf JSON. Just iterate through the pages, and for each page, remove any annotation whose `/Type` is `/Link`.

**Comment by Geo-Van** [comment]
Hello,
Very glad for the NEW BEHAVIOR (After version 11.8.0):
Applying { --flatten-annotations=all or --flatten-annotations=print or --flatten-annotations=screen } on a PDF file,
it will flattens ANNOTATIONS, but will KEEP the HYPERLINKS contained in the initial pdf file.

Please, let make a request for old Windows XP SP3 users:
The last  Windows XP SP3 qpdf working version is the:
qpdf-10.0.1-bin-msvc32  -->  need msvc32 version (NEEDED also the --> Visual C++ Redistributable for Visual Studio 2015).

Can you please create as a special request for XP users an update covering the above FIX (flattens ANNOTATIONS, but will KEEP the HYPERLINKS), lets say qpdf-10.0.1b  ?

We will be very happy if you can do it for us.
Thank you very much!

**Comment by jberkenbilt** [comment]
I'm sorry, I won't be able to create a fix for 10.0.1. The qpdf code has changed radically since then.

**Comment by Geo-Van** [comment]
Thank you for your reply,
Please let me kindly ask:
For that version 10.0.1, is there a "switch" we can use in order to flattens ANNOTATIONS, but will KEEP the HYPERLINKS?
Or is it impossible?

**Comment by jberkenbilt** [comment]
There's no switch. Sorry about that.

**Comment by Geo-Van** [comment]
I understand, thank you very much.
Please let me finally ask:
Is there a way to run the latest version qpdf-11.8.0 in windows XP, please? (Maybe by install any other software etc)?
If not, can you please make the qpdf-11.8.0 available to run in Windows XP?
We will appreciate it very much, thank you!

**Comment by jberkenbilt** [comment]
I'm sorry -- I appreciate your enthusiasm about this, but Windows XP has been out of support for a very long time (like 15 years), and there is just no way I am going to spend time doing anything to help make it possible to run modern qpdf on Windows XP.

**Comment by jberkenbilt** [comment]
@Geo-Van I'm trying to figure out if there's something I can do here. Do you know what happens if you run newer than 10.0.1 on XP? Looking at the fix for this bug, it was just one line of code, so backporting to 10.0.1 would be easy. I'm just not sure I can build qpdf that will work on XP. I notice that 10.0.1 was the last version of qpdf to use native crypto rather than openssl. If you set the environment variable `QPDF_CRYPTO_PROVIDER` to `native`, does the latest qpdf executable work on Windows XP? Have you tried all the distributions? The default distribution is msvc64, but you could try mingw32, which was the default at the time. See if the latest qpdf works with the native crypto provider and the mingw 32-bit build. You could also try the msvc 32-bit build. If you tell me what about the latest qpdf prevents it from running on Windows XP, I might have other ideas, though I'm far from a Windows expert. I could also try to send 10.0.1 with the one-line fix through CI and see if the result works, but I don't have any way to test or debug it.

**Comment by jberkenbilt** [comment]
Hmm, probably shouldn't have posted anything that might get your hopes up. I dusted off an old windows XP vm and see that the current qpdf executables won't run at all. This is probably based on the compiler toolchains, etc. Okay, I'm afraid I don't have time to look into it further. I don't think building qpdf 10.0.1 with current tools will work, and I don't have a good way to set up the environment that would have been in use at that time.

If anyone wanted to take a crack at this, this is the one line that changed: https://github.com/qpdf/qpdf/blob/78a7193fe2a5d14ca422407b35de9302f01d62ec/libqpdf/QPDFPageDocumentHelper.cc#L128

You can find the change here: https://github.com/qpdf/qpdf/commit/78a7193fe2a5d14ca422407b35de9302f01d62ec
by locating QPDFPageDocumentHelper.cc. The rest of the commit is documentation and tests.

It would be pretty easy to find that line in qpdf 10.0.1. I won't be identical because of reformatting, etc., but it should be there. In fact, here it is: https://github.com/qpdf/qpdf/blob/0c1122a729db2863b49bd4c4252a8997a0d4c787/libqpdf/QPDFPageDocumentHelper.cc#L171

So changing that one line and rebuilding using older tools would probably produce an executable that worked on Windows XP. Version 10.0.1 was released on April 9, 2020 and would have used whatever platform the `windows-latest` GitHub Actions runner would have been at that time. That runner may no longer exist, but it is always possible to build qpdf locally on Windows. 10.0.1 is before the cmake change, so getting the build environment set up was harder, but there is documentation in that version that describes how to do it.

I'm going to close this again, but I might be able to answer specific questions if someone attempts this.

**Comment by Geo-Van** [comment]
Hello,
Many thanks for trying to help regarding Windows Xp.
I really love your software, QPDF, and i wanted it to see it working also in XP regarding the correct behavior of flatten annotations without killing hyperlinks/links.
If i was a programmer, surely i would done what you advice in your previous post.
But unfortunately i understand nothing.
So, (forgive me for insisting) i will kindly ask if you can do it for us - not now, but anytime in future, we will be very very happy to see it working on XP with the correct behavior.
Best Regards
George
________________________________

**Comment by Geo-Van** [comment]
Hello dear jberkenbilt
I try a lot but no success making the correction on last XP version.... qpdf 10.0.1

There is a free software that can change executable .exe files known as ResourceHacker
Can be downloaded from:
https://www.softpedia.com/get/PORTABLE-SOFTWARE/Programming/Portable-Resource-Hacker.shtml

Can you please check if you can help us with this and create  a fix for the last XP working version of qpdf 10.0.1 ?

**Comment by jberkenbilt** [comment]
Doesn't sound relevant. The newer compiler toolchains don't create XP-compatible executables. You have to use an older toolchain, but it can't be so old that it doesn't support the current features in qpdf. The only way to do this that I know of would be to put together a system with the tools that were in use at the time of the 10.0.1 release, and I don't have an easy way to do that.
